File: scripts/proof-workers/orchestrator/cleanup.py
"""Session teardown and worktree removal."""

from __future__ import annotations

from pathlib import Path

from . import git, tmux
from .config import RunConfig
from .state import StateManager


def run_cleanup(cfg: RunConfig, keep_worktrees: bool = False) -> None:
    """Clean up tmux session, signal files, and optionally worktrees."""

    print("+" + "=" * 38 + "+")
    print("|  Orchestrator Cleanup                 |")
    print("+" + "=" * 38 + "+")
    print()

    # Kill tmux session
    if tmux.session_exists(cfg.tmux_session):
        print(f"Killing tmux session: {cfg.tmux_session}")
        tmux.kill_session(cfg.tmux_session)
        print("  Done.")
    else:
        print(f"No tmux session '{cfg.tmux_session}' found.")

    # Clean signal files
    print()
    print("Cleaning signal files...")
    for i in range(1, cfg.num_workers + 1):
        StateManager.signal_path(i).unlink(missing_ok=True)
    print("  Done.")

    # Remove worktrees
    if keep_worktrees:
        print()
        print("Keeping worktrees (--keep-worktrees specified).")
    else:
        print()
        print("Removing worktrees...")
        for name, repo_cfg in cfg.repos.items():
            wt_base = Path(repo_cfg.worktree_base)
            if not wt_base.is_dir():
                print(f"  No worktree directory for {name}: {wt_base}")
                continue

            for wt_dir in sorted(wt_base.glob("issue-*")):
                if wt_dir.is_dir():
                    print(f"  Removing: {wt_dir.name}")
                    success = git.remove_worktree(repo_cfg.path, str(wt_dir), force=True)
                    if not success:
                        print(f"    WARNING: Could not remove {wt_dir} (may need manual cleanup)")

            # Prune stale worktree references
            print(f"  Pruning stale worktree references for {name}...")
            git.prune_worktrees(repo_cfg.path)

            # Remove the worktree base dir if empty
            try:
                wt_base.rmdir()
            except OSError:
                pass  # Not empty, that's fine

    # Summary
    print()
    print("Cleanup complete.")
    print()
    for name, repo_cfg in cfg.repos.items():
        prefix = repo_cfg.branch_prefix
        if prefix:
            print(f"Branches are preserved for {name}. To list:")
            print(f"  git -C {repo_cfg.path} branch --list '{prefix}*'")
            print()
            print(f"To delete all branches with prefix '{prefix}':")
            print(f"  git -C {repo_cfg.path} branch --list '{prefix}*' | xargs -r git -C {repo_cfg.path} branch -D")
            print()

File: internal/orchestrator/cleanup.go
package orchestrator

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// RunCleanup stops worker processes, cleans signal files, and optionally removes worktrees.
func RunCleanup(cfg *RunConfig, keepWorktrees bool) {
	fmt.Println("+" + strings.Repeat("=", 38) + "+")
	fmt.Println("|  Orchestrator Cleanup                 |")
	fmt.Println("+" + strings.Repeat("=", 38) + "+")
	fmt.Println()

	// Stop all worker processes
	pm := GetProcessManager()
	runningWorkers := pm.GetRunningWorkers()
	if len(runningWorkers) > 0 {
		fmt.Printf("Stopping %d worker processes...\n", len(runningWorkers))
		pm.StopAll()
		fmt.Println("  Done.")
	} else {
		fmt.Println("No running worker processes found.")
	}

	// Clean signal files
	fmt.Println()
	fmt.Println("Cleaning signal files...")
	sm := NewStateManager(cfg)
	for i := 1; i <= cfg.NumWorkers; i++ {
		os.Remove(sm.SignalPath(i))
	}
	fmt.Println("  Done.")

	// Remove worktrees
	if keepWorktrees {
		fmt.Println()
		fmt.Println("Keeping worktrees (--keep-worktrees specified).")
	} else {
		fmt.Println()
		fmt.Println("Removing worktrees...")
		for name, repoCfg := range cfg.Repos {
			wtBase := repoCfg.WorktreeBase
			info, err := os.Stat(wtBase)
			if err != nil || !info.IsDir() {
				fmt.Printf("  No worktree directory for %s: %s\n", name, wtBase)
				continue
			}

			// Find all issue-* directories
			entries, _ := os.ReadDir(wtBase)
			var wtDirs []string
			for _, e := range entries {
				if e.IsDir() && strings.HasPrefix(e.Name(), "issue-") {
					wtDirs = append(wtDirs, filepath.Join(wtBase, e.Name()))
				}
			}
			sort.Strings(wtDirs)

			for _, wtDir := range wtDirs {
				fmt.Printf("  Removing: %s\n", filepath.Base(wtDir))
				if !RemoveWorktree(repoCfg.Path, wtDir, true) {
					fmt.Printf("    WARNING: Could not remove %s (may need manual cleanup)\n", wtDir)
				}
			}

			// Prune stale worktree references
			fmt.Printf("  Pruning stale worktree references for %s...\n", name)
			PruneWorktrees(repoCfg.Path)

			// Remove the worktree base dir if empty
			os.Remove(wtBase) // Will fail if not empty, that's fine
		}
	}

	// Summary
	fmt.Println()
	fmt.Println("Cleanup complete.")
	fmt.Println()
	for _, repoCfg := range cfg.Repos {
		prefix := repoCfg.BranchPrefix
		if prefix != "" {
			fmt.Printf("Branches are preserved for %s. To list:\n", repoCfg.Name)
			fmt.Printf("  git -C %s branch --list '%s*'\n", repoCfg.Path, prefix)
			fmt.Println()
			fmt.Printf("To delete all branches with prefix '%s':\n", prefix)
			fmt.Printf("  git -C %s branch --list '%s*' | xargs -r git -C %s branch -D\n", repoCfg.Path, prefix, repoCfg.Path)
			fmt.Println()
		}
	}
}
